src/ImageSharp/Formats/Icon/IconDecoderCore.cs

```diff
@@ -0,0 +1,207 @@
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        // Stream may not at 0.
+        long basePosition = stream.Position;
+        this.ReadHeader(stream);
+
+        Span<byte> flag = stackalloc byte[PngConstants.HeaderBytes.Length];
+
+        List<(Image<TPixel> Image, bool IsPng, int Index)> decodedEntries = new(this.Entries.Length);
+
+        for (int i = 0; i < this.Entries.Length; i++)
+        {
+            ref IconDirEntry entry = ref this.Entries[i];
+
+            // If we hit the end of the stream we should break.
+            if (stream.Seek(basePosition + entry.ImageOffset, SeekOrigin.Begin) >= stream.Length)
+            {
+                break;
+            }
+
+            // There should always be enough bytes for this regardless of the entry type.
+            if (stream.Read(flag) != PngConstants.HeaderBytes.Length)
+            {
+                break;
+            }
+
+            // Reset the stream position.
+            stream.Seek(-PngConstants.HeaderBytes.Length, SeekOrigin.Current);
+
+            bool isPng = flag.SequenceEqual(PngConstants.HeaderBytes);
+
+            // Decode the frame into a temp image buffer. This is disposed after the frame is copied to the result.
+            Image<TPixel> temp = this.GetDecoder(isPng).Decode<TPixel>(stream, cancellationToken);
+            decodedEntries.Add((temp, isPng, i));
+
+            // Since Windows Vista, the size of an image is determined from the BITMAPINFOHEADER structure or PNG image data
+            // which technically allows storing icons with larger than 256 pixels, but such larger sizes are not recommended by Microsoft.
+            this.Dimensions = new(Math.Max(this.Dimensions.Width, temp.Size.Width), Math.Max(this.Dimensions.Height, temp.Size.Height));
+        }
+
+        ImageMetadata metadata = new();
+        BmpMetadata? bmpMetadata = null;
+        PngMetadata? pngMetadata = null;
+        Image<TPixel> result = new(this.Options.Configuration, metadata, decodedEntries.Select(x =>
+        {
+            ImageFrame<TPixel> target = new(this.Options.Configuration, this.Dimensions);
+            ImageFrame<TPixel> source = x.Image.Frames.RootFrameUnsafe;
+            for (int y = 0; y < source.Height; y++)
+            {
+                source.PixelBuffer.DangerousGetRowSpan(y).CopyTo(target.PixelBuffer.DangerousGetRowSpan(y));
+            }
+
+            // Copy the format specific frame metadata to the image.
+            if (x.IsPng)
+            {
+                if (x.Index == 0)
+                {
+                    pngMetadata = x.Image.Metadata.GetPngMetadata();
+                }
+
+                // Bmp does not contain frame specific metadata.
+                target.Metadata.SetFormatMetadata(PngFormat.Instance, target.Metadata.GetPngFrameMetadata());
+            }
+            else if (x.Index == 0)
+            {
+                bmpMetadata = x.Image.Metadata.GetBmpMetadata();
+            }
```

Comment:
I think the BitsPerPixel property is important and will be used by the encoder.
Such as the flutter.ico
It has 4-bit 8-bit and 32-bit images of the same size.
We must tell the encoder to encode using the same BitsPerPixel.

Reply:
No we’d tell the encoder to encode each frame independently. Your ico/cut frame metadata should contain the bits per pixel property